src/Evolution/Systems/Cce/WorldtubeInterfaceManager.hpp

```diff
@@ -0,0 +1,144 @@
+ * `std::unique_ptr<GHWorldtubeInterfaceManager>` with cloned state.
+ * - `GHWorldtubeInterfaceManager::insert_gh_data()`: should store the portions
+ * of the provided generalized harmonic data that are required to provide useful
+ * boundary values for the CCE evolution at requested timesteps.
+ * - `GHWorldtubeInterfaceManager::request_gh_data()`: should register requests
+ * from the CCE evolution for boundary data.
+ * - `GHWorldtubeInterfaceManager::try_retrieve_gh_data()`: should return a
+ * `std::tuple<TimeStepId, tnsr::aa<DataVector, 3>,
+ *  tnsr::iaa<DataVector, 3>, tnsr::aa<DataVector, 3>>` containing the boundary
+ * data associated with the least recent requested timestep if enough data has
+ * been supplied via `insert_gh_data()` to determine the boundary data.
+ * Otherwise, return a `boost::none` to indicate that the CCE system must
+ * continue waiting for generalized harmonic input.
+ */
+class GHWorldtubeInterfaceManager : public PUP::able {
+ public:
+  using creatable_classes = tmpl::list<GHLockstepInterfaceManager>;
+
+  WRAPPED_PUPable_abstract(GHWorldtubeInterfaceManager);  // NOLINT
+
+  virtual std::unique_ptr<GHWorldtubeInterfaceManager> get_clone() const
+      noexcept = 0;
+
+  virtual void insert_gh_data(
+      const TimeStepId& time_id,
+      const tnsr::aa<DataVector, 3>& spacetime_metric,
+      const tnsr::iaa<DataVector, 3>& phi, const tnsr::aa<DataVector, 3>& pi,
+      const tnsr::aa<DataVector, 3>& dt_spacetime_metric,
+      const tnsr::iaa<DataVector, 3>& dt_phi,
+      const tnsr::aa<DataVector, 3>& dt_pi) noexcept = 0;
+
+  virtual void request_gh_data(const TimeStepId&) noexcept = 0;
+
+  virtual auto try_retrieve_first_ready_gh_data() noexcept -> boost::optional<
+      std::tuple<TimeStepId, tnsr::aa<DataVector, 3>, tnsr::iaa<DataVector, 3>,
+                 tnsr::aa<DataVector, 3>>> = 0;
+};
+
+/*!
+ * \brief Simple implementation of a `GHWorldtubeInterfaceManager` that only
+ * provides boundary data on matching `TimeStepId`s
+ *
+ * \details This version of the interface manager assumes that the CCE system
+ * and the generalized harmonic system that it communicates with evolve with an
+ * identical time stepper and on identical time step intervals (they evolve in
+ * 'lock step'). As a result, and to streamline communications, new data is
+ * always immediately 'ready' and requests are ignored to produce the behavior
+ * of just immediately gauge-transforming and sending the data to the CCE
+ * component as soon as it becomes available from the GH system.
+ *
+ * \warning Using this interface manager when the GH components and the CCE
+ * evolution are not identically stepped is considered undefined behavior. The
+ * outcome will likely be that CCE will fail to evolve and the boundary data
+ * will be continually inserted into a rapidly expanding inbox.
+ */
+class GHLockstepInterfaceManager : public GHWorldtubeInterfaceManager {
```

Comment:
Please separate the classes, and put them into a subdirectory of Cce as that directory is getting very full...